### PR TITLE
KAFKA-17036: KIP-919 supports for createAcls, deleteAcls, describeAcls

### PR DESCRIPTION
**Pull request by FrankYang0529**
* Use `LeastLoadedBrokerOrActiveKController` for `describeAcls`, `createAcls`, and `deleteAcls` APIs.
* Add `--bootstrap-controller` to `AclCommand`.
* Add related integration tests to `AclCommandTest`.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: tools/src/test/java/org/apache/kafka/tools/ClusterToolTest.java [hunk 1]
```diff
@@ -74,6 +74,7 @@ public void testUnregister(ClusterInstance clusterInstance) {
 
     @ClusterTest(types = {Type.KRAFT, Type.CO_KRAFT})
     public void testClusterIdWithBootstrapController(ClusterInstance clusterInstance) {
+        System.out.println("bootstrapControllers: " + clusterInstance.bootstrapControllers());
```
Comment (chia7712) [inline tools/src/test/java/org/apache/kafka/tools/ClusterToolTest.java]: what is this?

Comment (FrankYang0529) [reply tools/src/test/java/org/apache/kafka/tools/ClusterToolTest.java]: Sorry, this is my local test. Removed it. Thank you.

File: core/src/main/scala/kafka/admin/AclCommand.scala [hunk 2]
```diff
@@ -626,18 +637,22 @@ object AclCommand extends Logging {
...
-        CommandLineUtils.printUsageAndExit(parser, "Only one of --bootstrap-server or --authorizer must be specified")
+      if (options.has(bootstrapServerOpt) && options.has(bootstrapControllerOpt))
+        CommandLineUtils.printUsageAndExit(parser, "Only one of --bootstrap-server or --bootstrap-controller must be specified")
+
+      val hasServerOrController = options.has(bootstrapServerOpt) || options.has(bootstrapControllerOpt)
+      if (hasServerOrController && options.has(authorizerOpt))
+        CommandLineUtils.printUsageAndExit(parser, "The --authorizer option can only be used without --bootstrap-server or --bootstrap-controller")
 
-      if (!options.has(bootstrapServerOpt)) {
+      if (!options.has(bootstrapServerOpt) && !options.has(bootstrapControllerOpt)) {
```
Comment (chia7712) [inline core/src/main/scala/kafka/admin/AclCommand.scala]: `!options.has(bootstrapServerOpt) && !options.has(bootstrapControllerOpt)`  can be replaced by `hasServerOrController`

File: core/src/main/scala/kafka/admin/AclCommand.scala [hunk 3]
```diff
@@ -626,18 +637,22 @@ object AclCommand extends Logging {
...
         CommandLineUtils.checkRequiredArgs(parser, options, authorizerPropertiesOpt)
         System.err.println(AclCommand.AuthorizerDeprecationMessage)
       }
 
-      if (options.has(commandConfigOpt) && !options.has(bootstrapServerOpt))
-        CommandLineUtils.printUsageAndExit(parser, "The --command-config option can only be used with --bootstrap-server option")
+      if (options.has(commandConfigOpt) && (!options.has(bootstrapServerOpt) && !options.has(bootstrapControllerOpt)))
```
Comment (chia7712) [inline core/src/main/scala/kafka/admin/AclCommand.scala]: ditto

File: core/src/main/scala/kafka/admin/AclCommand.scala [hunk 4]
```diff
@@ -626,18 +637,22 @@ object AclCommand extends Logging {
...
+        CommandLineUtils.printUsageAndExit(parser, "The --command-config option can only be used with --bootstrap-server or --bootstrap-controller option")
 
-      if (options.has(authorizerPropertiesOpt) && options.has(bootstrapServerOpt))
+      if (options.has(authorizerPropertiesOpt) && (options.has(bootstrapServerOpt) || options.has(bootstrapControllerOpt)))
```
Comment (chia7712) [inline core/src/main/scala/kafka/admin/AclCommand.scala]: ditto

**Review by chia7712 — COMMENTED** [review]
@FrankYang0529 thanks for this patch. Could you please make `testAclOperations`  run with controller also?

https://github.com/apache/kafka/blob/42f267a853734effc7aff24e30ec812aba84f834/core/src/test/scala/integration/kafka/api/PlaintextAdminIntegrationTest.scala#L1121

**Comment by FrankYang0529** [comment]
Hi @chia7712, I addressed all comments. Thanks for your review.

**Comment by chia7712** [comment]
@FrankYang0529 Could you please move all tests to `BootstrapControllersIntegrationTest` after #16822 gets merged?

**Comment by chia7712** [comment]
@FrankYang0529 any updates?

**Comment by FrankYang0529** [comment]
> @FrankYang0529 Could you please move all tests to `BootstrapControllersIntegrationTest` after #16822 gets merged?

Hi @chia7712, updated it. Could you take a look? Thank you.

File: core/src/test/java/kafka/test/server/BootstrapControllersIntegrationTest.java [hunk 5]
```diff
@@ -235,4 +254,48 @@ private void testIncrementalAlterConfigs(ClusterInstance clusterInstance, boolea
...
+            AccessControlEntry accessControlEntry = new AccessControlEntry("User:ANONYMOUS", "*", AclOperation.DESCRIBE, AclPermissionType.ALLOW);
+            AclBinding aclBinding = new AclBinding(resourcePattern, accessControlEntry);
+            assertDoesNotThrow(() -> admin.createAcls(Collections.singleton(aclBinding)).all().get(1, TimeUnit.MINUTES));
+
+            List<Authorizer> allAuthorizers = new ArrayList<>();
+            allAuthorizers.addAll(clusterInstance.brokers().values().stream()
+                    .map(server -> server.authorizer().get()).collect(Collectors.toList()));
+            allAuthorizers.addAll(clusterInstance.controllers().values().stream()
+                    .map(server -> server.authorizer().get()).collect(Collectors.toList()));
+            allAuthorizers.forEach(authorizer -> kafka.utils.TestUtils.waitAndVerifyAcls(
```
Comment (chia7712) [inline core/src/test/java/kafka/test/server/BootstrapControllersIntegrationTest.java]: Could you please add those code to `ClusterInstance` as a helper method?

**Review by chia7712 — COMMENTED** [review]
@FrankYang0529 thanks for this nice patch. one one small comment is left

File: core/src/test/java/kafka/test/server/BootstrapControllersIntegrationTest.java (hunk 5 above)
Comment (FrankYang0529) [reply core/src/test/java/kafka/test/server/BootstrapControllersIntegrationTest.java]: Yes, add a method `authorizers` to `ClusterInstance`. Thanks.

File: core/src/test/java/kafka/test/ClusterInstance.java [hunk 6]
```diff
@@ -204,4 +207,13 @@ default void waitForTopic(String topic, int partitions) throws InterruptedExcept
                 60000L, "Timeout waiting for controller metadata propagating to brokers");
         }
     }
+
+    default List<Authorizer> authorizers() {
```
Comment (chia7712) [inline core/src/test/java/kafka/test/ClusterInstance.java:218]: Could you please check the existence of `Option` ?

File: core/src/test/java/kafka/test/ClusterInstance.java [hunk 7]
```diff
@@ -204,4 +207,13 @@ default void waitForTopic(String topic, int partitions) throws InterruptedExcept
                 60000L, "Timeout waiting for controller metadata propagating to brokers");
         }
     }
+
```
Comment (chia7712) [inline core/src/test/java/kafka/test/ClusterInstance.java:217]: for another, please add `waitAcls`. for example:

```java
    default void waitAcls(AclBindingFilter filter, Collection<AccessControlEntry> entries) throws InterruptedException {
        for (Authorizer authorizer : authorizers()) {
            TestUtils.waitForCondition(() -> {
                Assertions.assertIterableEquals(authorizer.acls(filter), entries);
                return true;
            }, "except: ");
        }
    }
```

**Review by chia7712 — COMMENTED** [review]
@FrankYang0529 thanks for updated PR

File: core/src/main/scala/kafka/admin/AclCommand.scala [hunk 8]
```diff
@@ -626,18 +637,22 @@ object AclCommand extends Logging {
...
         System.err.println(AclCommand.AuthorizerDeprecationMessage)
       }
 
-      if (options.has(commandConfigOpt) && !options.has(bootstrapServerOpt))
-        CommandLineUtils.printUsageAndExit(parser, "The --command-config option can only be used with --bootstrap-server option")
+      if (options.has(commandConfigOpt) && (!hasServerOrController))
+        CommandLineUtils.printUsageAndExit(parser, "The --command-config option can only be used with --bootstrap-server or --bootstrap-controller option")
 
-      if (options.has(authorizerPropertiesOpt) && options.has(bootstrapServerOpt))
+      if (options.has(authorizerPropertiesOpt) && (hasServerOrController))
```
Comment (chia7712) [inline core/src/main/scala/kafka/admin/AclCommand.scala]: nit: `if (options.has(authorizerPropertiesOpt) && hasServerOrController)`

File: core/src/test/java/kafka/test/ClusterInstance.java [hunk 9]
```diff
@@ -204,4 +210,28 @@ default void waitForTopic(String topic, int partitions) throws InterruptedExcept
...
+    default void waitAcls(AclBindingFilter filter, Collection<AccessControlEntry> entries) throws InterruptedException {
+        for (Authorizer authorizer : authorizers()) {
+            AtomicReference<Set<AccessControlEntry>> actualEntries = new AtomicReference<>(new HashSet<>());
+            TestUtils.waitForCondition(() -> {
+                Set<AccessControlEntry> accessControlEntrySet = new HashSet<>();
+                authorizer.acls(filter).forEach(aclBinding -> accessControlEntrySet.add(aclBinding.entry()));
+                actualEntries.set(accessControlEntrySet);
+                return accessControlEntrySet.containsAll(entries) && entries.containsAll(accessControlEntrySet);
+            }, () -> "except acls: " + entries + ", actual acls: " + actualEntries.get());
```
Comment (chia7712) [inline core/src/test/java/kafka/test/ClusterInstance.java]: `expected acls:

**Review by chia7712 — COMMENTED** [review]
@FrankYang0529 thanks for this patch

**Comment by FrankYang0529** [comment]
Hi @chia7712, thanks for the review. I address all comments.

**Comment by chia7712** [comment]
@FrankYang0529 could you please fix conflicts?

**Comment by FrankYang0529** [comment]
> @FrankYang0529 could you please fix conflicts?

Fixed it. Thank you.

File: core/src/test/java/kafka/admin/AclCommandTest.java [hunk 10]
```diff
@@ -261,6 +297,27 @@ public void testAclCliWithClientId(ClusterInstance cluster) {
         appender.setClassLogger(AppInfoParser.class, Level.WARN);
         try {
             testAclCli(cluster, adminArgs(cluster.bootstrapServers(), Optional.of(TestUtils.tempFile("client.id=my-client"))));
+        } catch (InterruptedException | IOException e) {
```
Comment (chia7712) [inline core/src/test/java/kafka/admin/AclCommandTest.java]: Could you please do a bit refactor for those tests?
```java
        try (LogCaptureAppender appender = LogCaptureAppender.createAndRegister()) {
            appender.setClassLogger(AppInfoParser.class, Level.WARN);
            testAclCli(cluster, adminArgs(cluster.bootstrapServers(), Optional.of(TestUtils.tempFile("client.id=my-client"))));
            assertEquals(0, appender.getEvents().stream()
                    .filter(e -> e.getLevel().equals(Level.WARN.toString()))
                    .filter(e -> e.getThrowableClassName().filter(name -> name.equals(InstanceAlreadyExistsException.class.getName())).isPresent())
                    .count(), "There should be no warnings about multiple registration of mbeans");
        }
```

File: core/src/test/java/kafka/admin/AclCommandTest.java [hunk 11]
```diff
@@ -261,6 +297,27 @@ public void testAclCliWithClientId(ClusterInstance cluster) {
...
+
+    @ClusterTest(types = {Type.KRAFT}, serverProperties = {
+            @ClusterConfigProperty(key = AUTHORIZER_CLASS_NAME_CONFIG, value = STANDARD_AUTHORIZER)
+    })
+    public void testAclCliWithClientIdAndBootstrapController(ClusterInstance cluster) {
+        LogCaptureAppender appender = LogCaptureAppender.createAndRegister();
+        appender.setClassLogger(AppInfoParser.class, Level.WARN);
+        try {
+            testAclCli(cluster, adminArgsWithBootstrapController(cluster.bootstrapControllers(), Optional.of(TestUtils.tempFile("client.id=my-client"))));
+        } catch (InterruptedException | IOException e) {
```
Comment (chia7712) [inline core/src/test/java/kafka/admin/AclCommandTest.java]: ditto

**Review by chia7712 — COMMENTED** [review]
@FrankYang0529 thanks for this patch.

**Comment by FrankYang0529** [comment]
Hi @chia7712, I addressed all comments. Could you take a look again? Thank you.

File: core/src/test/java/kafka/admin/AclCommandTest.java [hunk 12]
```diff
@@ -246,32 +275,54 @@ public void testProducerConsumerCliWithAuthorizer(ClusterInstance cluster) {
...
+        try (LogCaptureAppender appender = LogCaptureAppender.createAndRegister()) {
+            appender.setClassLogger(AppInfoParser.class, Level.WARN);
             testAclCli(cluster, adminArgs(cluster.bootstrapServers(), Optional.of(TestUtils.tempFile("client.id=my-client"))));
-        } finally {
-            appender.close();
+            assertEquals(0, appender.getEvents().stream()
+                    .filter(e -> e.getLevel().equals(Level.WARN.toString()))
+                    .filter(e -> e.getThrowableClassName().filter(name -> name.equals(InstanceAlreadyExistsException.class.getName())).isPresent())
+                    .count(), "There should be no warnings about multiple registration of mbeans");
+        } catch (InterruptedException | IOException e) {
```
Comment (chia7712) [inline core/src/test/java/kafka/admin/AclCommandTest.java]: we don't need to catch the exception. Instead, we can add the exception to the method signature.

File: core/src/test/java/kafka/admin/AclCommandTest.java [hunk 13]
```diff
@@ -246,32 +275,54 @@ public void testProducerConsumerCliWithAuthorizer(ClusterInstance cluster) {
...
+    })
+    public void testAclCliWithClientIdAndBootstrapController(ClusterInstance cluster) {
+        try (LogCaptureAppender appender = LogCaptureAppender.createAndRegister()) {
+            appender.setClassLogger(AppInfoParser.class, Level.WARN);
+            testAclCli(cluster, adminArgsWithBootstrapController(cluster.bootstrapControllers(), Optional.of(TestUtils.tempFile("client.id=my-client"))));
+            assertEquals(0, appender.getEvents().stream()
+                    .filter(e -> e.getLevel().equals(Level.WARN.toString()))
+                    .filter(e -> e.getThrowableClassName().filter(name -> name.equals(InstanceAlreadyExistsException.class.getName())).isPresent())
+                    .count(), "There should be no warnings about multiple registration of mbeans");
+        } catch (InterruptedException | IOException e) {
```
Comment (chia7712) [inline core/src/test/java/kafka/admin/AclCommandTest.java]: ditto

**Review by chia7712 — COMMENTED** [review]
@FrankYang0529 overall LGTM. two small comments are left.

File: core/src/test/java/kafka/admin/AclCommandTest.java (hunk 12 above)
Comment (FrankYang0529) [reply core/src/test/java/kafka/admin/AclCommandTest.java]: Hi @chia7712, thanks for your review. I add exceptions to the method signature.

**Comment by FrankYang0529** [comment]
Hi @chia7712, could you help me review this when you have time? Thank you.

**Review by chia7712 — APPROVED** [review]

**Pull request merged**